### ISSUE TITLE
Bug: cleanExpiredRevision may delete current revision

### ISSUE THREAD

**Issue by dbug-dk** [KusionStack/operating#148]
### Minimal reproduce step

1. 创建一个collaset，当前为revision1
2. 更新一个不存在的镜像，模拟发布失败(当前inplaceUpdate和recreate的方式会立即成功,但需要考虑以后pod原地升级不成功的场景)， 会创建revision2.
3. 再次更新一个不同于revision1的新镜像，会创建revision3.
这时候，由于第二步里发布失败了，current revision = revision1， update revision = revision3

如果collaset的history limit 默认为0，只会额外保留两个revision，删除1个revision时，根据排序，会直接删除最老的revision，也就是当前的revision1. 

影响:  代码会panic
![image](https://github.com/KusionStack/operating/assets/7811716/c0377e24-eef2-467f-a526-cae489088f14)

### What did you expect to see?

清理controllerRevision时，过滤current revision和update revision

### What did you see instead

当前清理controllerRevision时，只会优先删除老版本的revision
代码里判断
![image](https://github.com/KusionStack/operating/assets/7811716/56fb598c-dd4f-48a1-ba3a-c58befcf6ec2)

![image](https://github.com/KusionStack/operating/assets/7811716/47481fdf-bf51-4205-a6d2-b51b7e66359e)

### What is your KusionStack components and its version?

main

**Comment by ColdsteelRail** [comment]
I'm working on this issue, could you please assign me this issue? thx

**Comment by ColdsteelRail** [comment]
Hello, I'm resolving this problem, could you please assign me this issue? lots of thanks

**Comment by ColdsteelRail** [comment]
Hello, have you resolved this problem?
If not, could you please assign this issue to me
lots of thanks

**Comment by ColdsteelRail** [comment]
Hello, are you resolving this problem?
Assign request if not, lots of thanks

**Comment by ColdsteelRail** [comment]
Hello, are you resolving this problem right now?
Assign request

**Comment by ColdsteelRail** [comment]
The same case occurs in the official proposal [revision-consistency](https://www.kusionstack.io/docs/operating/manuals/collaset#revision-consistency) : 
- Minimal Reproduce:
1. follow the step1, step2, step3 to provision 3 pods with revision 1,2 and 3 respectively
2. **delete the pod with revision 1 (nginx:1.25.2)**
- What is expected:
a pod with revision 1 (nginx:1.25.2) should be recreated

- What happened instead:
a pod with revision 2 (nginx:1.23.4) is recreated instead

![pod-revision-bug](https://github.com/KusionStack/operating/assets/42696362/8fc376ac-d44e-445f-b8d8-a791247bbe7b)

**Comment by ColdsteelRail** [comment]
Two potential proposals for this issue, however, with compromises:
1. Filter the current and update revisions in ```cleanExpiredRevision()```, this can maintain [revision-consistency](https://www.kusionstack.io/docs/operating/manuals/collaset#revision-consistency) but destroy the natural of ```spec.HistoryLimit```
2. Make sure ```podInfo.CurrentRevision``` is always not nil by modifying the below ```attachPodUpdateInfo()```, this can maintain the natural of ```spec.HistoryLimit``` but destroy [revision-consistency](https://www.kusionstack.io/docs/operating/manuals/collaset#revision-consistency).

![currentRevisionNil](https://github.com/KusionStack/operating/assets/42696362/6c79340b-a019-4086-b913-27903535ff01)


There can be more considerable solutions, could you please help me check it out?
cc @dbug-dk

**Comment by ColdsteelRail** [comment]
#### 提供一个可以100%复现bug的示例，便于进一步修复：
1. 创建replicas=3的collaset
```
apiVersion: apps.kusionstack.io/v1alpha1
kind: CollaSet
metadata:
  name: foo
spec:
  replicas: 3 # indicate the number of Pods to manage
  selector:
    matchLabels:
      app: foo
  updateStrategy:
    podUpgradePolicy: InPlaceIfPossible
  template:
    metadata:
      labels:
        app: foo
    spec:
      containers:
        - image: nginx:1.25.2 # exist
          name: nginx
```
2. 第一次修改镜像，同时设置partition=2
```
...
spec:
  updateStrategy:
    podUpgradePolicy: InPlaceIfPossible
    rollingUpdate:
      byPartition:
        partition: 2
  ...
    spec:
      containers:
        - image: nginx:1.24.0 # exist
          name: nginx
```
3. 第二次修改镜像，同时设置partition=1
```
...
spec:
  updateStrategy:
    podUpgradePolicy: InPlaceIfPossible
    rollingUpdate:
      byPartition:
        partition: 1
  ...
    spec:
      containers:
        - image: nginx:1.23.4 # exist
          name: nginx
```
4. 修改partition=3，更新全部副本，此时controller出现panic
```
...
spec:
  updateStrategy:
    podUpgradePolicy: InPlaceIfPossible
    rollingUpdate:
      byPartition:
        partition: 3
  ...
    spec:
      containers:
        - image: nginx:1.23.4 # exist
          name: nginx
```

#### 关键日志
```
I0121 03:04:21.431522       1 controller.go:220] controller/collaset-controller "msg"="Starting workers"  "worker count"=5
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x118 pc=0x14e9db3]

goroutine 377 [running]:
kusionstack.io/operating/pkg/controllers/collaset/utils.GetPodRevisionPatch(0x0)
	/workspace/pkg/controllers/collaset/utils/pod.go:95 +0x33
kusionstack.io/operating/pkg/controllers/collaset/utils.ApplyPatchFromRevision(0xc00047f690, 0x0?)
	/workspace/pkg/controllers/collaset/utils/pod.go:106 +0x36
kusionstack.io/operating/pkg/controllers/collaset/utils.GetPodFromRevision(...)
	/workspace/pkg/controllers/collaset/utils/pod.go:158
kusionstack.io/operating/pkg/controllers/collaset/utils.NewPodFrom({0x1bd5a20, 0xc0000e8000}, 0xc0003bcd20, 0x0, {0xc00047fb30, 0x1, 0x1942b9d?})
	/workspace/pkg/controllers/collaset/utils/pod.go:73 +0xa5
kusionstack.io/operating/pkg/controllers/collaset/synccontrol.(*inPlaceIfPossibleUpdater).AnalyseAndGetUpdatedPod(0xc0005ac720, 0xc00016d600, 0xc0008d3e40)
	/workspace/pkg/controllers/collaset/synccontrol/update.go:230 +0x159
kusionstack.io/operating/pkg/controllers/collaset/synccontrol.(*RealSyncControl).Update.func3(0xc000681738?, {0xc0006817d0?, 0x11a4325?})
	/workspace/pkg/controllers/collaset/synccontrol/sync_control.go:526 +0xc3
kusionstack.io/operating/pkg/controllers/utils.SlowStartBatch.func1(0xc0006817b8?)
	/workspace/pkg/controllers/utils/slow_start.go:57 +0x74
created by kusionstack.io/operating/pkg/controllers/utils.SlowStartBatch
	/workspace/pkg/controllers/utils/slow_start.go:55 +0x1dc
```

**Comment by wu8685** [comment]
fixed by #154
